keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2165 @@
+re-implement) the current runtime classes.
+
+One benefit of using a QoS-class resource to represent runtime classes would be
+the "automatic" visibility to what runtime classes are available on each node.
+E.g. adding a new runtime class on a node by re-configuration of the runtime
+would be immediately reflected in the node status without any additional admin
+tasks (like node labeling or manual creation of RuntimeClass objects).
+
+###### Splitting Pod QoS Class
+
+Currently the Pod QoS class is an implicit property of a Pod, tied to how the
+resource requests and limits of its containers are specified. QoS-class
+resources would make it possible for users to explicitly specify the Pod QoS
+class in the PodSpec, making it possible e.g. to create a guaranteed pod with a
+container whose memory limit is higher than the requests.
+
+Taking this idea further, QoS-class resources would also make it possible to
+split several properties implicit in the Pod QoS class (like eviction behavior
+or OOM scoring) into separate properties. For example, in the case of OOM
+scoring make it possible for the user to specify a burstable pod with low OOM
+priority (low chance of being killed).
+
+###### Pod priority class
+
+One wild idea would be to implement pod priority class mechanism (or part of
+it) as a QoS-class resource. This would be a
+[Kubernetes-managed](#kubernetes-managed-qos-class-resources) pod-level
+QoS-class resource. Likely benefits of using the QoS-class resources mechanism
+would be to be able to set per-namespace defaults with LimitRanges and allow
+permission-control to high-priority classes with ResourceQuotas.
+
+### Notes/Constraints/Caveats (Optional)
+
+The proposal only describes the details of implementation Phase 1 and some
+parts like permission control are not covered. The
+risk in this piecemeal approach is finding devil in the details, resulting in
+inconsistent and/or crippled and/or cumbersome end result. However, the missing
+parts are already outlined in [future work](#future-work) and there is aa lot
+of experience in extending the API and understanding which sort of solutions
+are functional and practical, mitigating the problem.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The identified risks are mainly related to usability or unfriendly users
+hogging all available high-priority QoS.
+
+- User assigning container to “unauthorized” class, causing interference and
+  access to unwanted set/amount of resources. This will be addressed in future
+  KEP introducing permission controls.
+- Cross-registering pod and container-level QoS-class resources: it would be
```

Comment:
_This is an informational comment and doesn't block merging the KEP._

Mitigation for this could be an API type, similar to GatewayClass or IngressClass, to define what
QoS assignments are possible and whether these are container-level, Pod-level, or support both.

Reply:
Maybe. But I think this creates another caveat: misconfiguration of this API object. Mis-configure the API object as pod-level resource when it should've been container-level (or vice versa) which would then prevent registration of the resource. And the risk for this is bigger (misconfiguration) than cross-registering which would be sort of mis-implementation on the runtime side

Reply:
Once this merges, I can open a PR specifically about how I'd deal with this.